### DISCUSSION TITLE
Trouble reproducing env for BERT pretraining

### DISCUSSION THREAD

**Discussion by azayz** [NVIDIA-NeMo/NeMo#2732] [Q&A] [answered]
Hey NeMo team, I'm having trouble reproducing nemo env for pretraining BERT on 4 GPUs, can you please point me to NeMo stable version that allows me to follow this tutorial https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/zh/latest/nlp/bert_pretraining.html . If you can provide commands swell that would be great or any other way to retrain BERT on multi GPUs. Thanks a lot in advance

**Comment by titu1994** [comment]
@ekmb or @yzhang123 would you have scripts for BERT pretraining ?

**Reply by azayz** [reply, marked as answer]
Nevermind I finally made it work using the following script NeMo/examples/nlp/language_modeling/bert_pretraining.py and the following  config : 

Python 3.9.6

and pip install Cython

then install these

absl-py==0.13.0
aiohttp==3.7.4.post0
alabaster==0.7.12
antlr4-python3-runtime==4.8
appdirs==1.4.4
argon2-cffi==20.1.0
async-timeout==3.0.1
attrdict==2.0.1
attrs==21.2.0
audioread==2.1.9
Babel==2.9.1
backcall==0.2.0
black==19.10b0
bleach==4.1.0
boto3==1.18.29
botocore==1.21.29
braceexpand==0.1.7
cachetools==4.2.2
certifi==2021.5.30
cffi==1.14.6
chardet==4.0.0
charset-normalizer==2.0.4
click==7.1.2
colorama==0.4.4
configparser==5.0.2
cycler==0.10.0
Cython==0.29.24
debugpy==1.4.1
decorator==5.0.9
defusedxml==0.7.1
Distance==0.1.3
docker-pycreds==0.4.0
docopt==0.6.2
docutils==0.17.1
editdistance==0.5.3
entrypoints==0.3
filelock==3.0.12
frozendict==2.0.6
fsspec==2021.7.0
future==0.18.2
g2p-en==2.1.0
gdown==3.13.0
gitdb==4.0.7
GitPython==3.1.18
google-auth==1.35.0
google-auth-oauthlib==0.4.5
grpcio==1.39.0
h5py==3.4.0
huggingface-hub==0.0.12
hydra-core==1.1.1
idna==3.2
imagesize==1.2.0
inflect==5.3.0
iniconfig==1.1.1
ipadic==1.0.0
ipykernel==6.2.0
ipython==7.26.0
ipython-genutils==0.2.0
ipywidgets==7.6.3
isort==4.3.21
jedi==0.18.0
jieba==0.42.1
Jinja2==3.0.1
jmespath==0.10.0
joblib==1.0.1
jsonschema==3.2.0
jupyter-client==7.0.1
jupyter-core==4.7.1
jupyterlab-pygments==0.1.2
jupyterlab-widgets==1.0.0
kaldi-io==0.9.4
kaldi-python-io==1.2.2
kaldiio==2.17.2
kiwisolver==1.3.1
latexcodec==2.0.1
librosa==0.8.1
llvmlite==0.37.0
Markdown==3.3.4
MarkupSafe==2.0.1
marshmallow==3.13.0
matplotlib==3.4.3
matplotlib-inline==0.1.2
mecab-python3==1.0.3
megatron-lm==2.2.0
mistune==0.8.4
mpmath==1.2.1
multidict==5.1.0
nbclient==0.5.4
nbconvert==6.1.0
nbformat==5.1.3
nemo-toolkit==1.2.0
nest-asyncio==1.5.1
nltk==3.6.2
notebook==6.4.3
numba==0.54.0
numpy==1.20.3
oauthlib==3.1.1
omegaconf==2.1.1
onnx==1.10.1
OpenCC==1.1.1.post1
packaging==21.0
pandas==1.3.2
pandocfilters==1.4.3
pangu==4.0.6.1
parameterized==0.8.1
parso==0.8.2
pathspec==0.9.0
pathtools==0.1.2
pesq==0.0.3
pexpect==4.8.0
pickleshare==0.7.5
Pillow==8.3.1
pip-api==0.0.20
pipreqs==0.4.10
pluggy==0.13.1
pooch==1.5.1
portalocker==2.3.1
prometheus-client==0.11.0
promise==2.3
prompt-toolkit==3.0.20
protobuf==3.17.3
psutil==5.8.0
ptyprocess==0.7.0
py==1.10.0
pyannote.core==4.1
pyannote.database==4.1.1
pyannote.metrics==3.0.1
pyasn1==0.4.8
pyasn1-modules==0.2.8
pybind11==2.7.1
pybtex==0.24.0
pybtex-docutils==1.0.1
pycparser==2.20
pyDeprecate==0.3.0
pydub==0.25.1
Pygments==2.10.0
pynini 
pyparsing==2.4.7
pypinyin==0.42.0
pyrsistent==0.18.0
PySocks==1.7.1
pystoi==0.3.3
pytest==6.2.4
pytest-runner==5.3.1
python-dateutil==2.8.2
pytorch-lightning==1.3.8
pytz==2021.1
PyYAML==5.4.1
pyzmq==22.2.1
rapidfuzz==1.5.0
regex==2021.8.21
requests==2.26.0
requests-oauthlib==1.3.0
resampy==0.2.2
rsa==4.7.2
ruamel.yaml==0.17.14
ruamel.yaml.clib==0.2.6
s3transfer==0.5.0
sacrebleu==2.0.0
sacremoses==0.0.45
scikit-learn==0.24.2
scipy==1.7.1
Send2Trash==1.8.0
sentencepiece==0.1.96
sentry-sdk==1.3.1
shellingham==1.4.0
shortuuid==1.0.1
simplejson==3.17.5
six==1.16.0
smmap==4.0.0
sniffio==1.2.0
snowballstemmer==2.1.0
sortedcontainers==2.4.0
SoundFile==0.10.3.post1
sox==1.4.1
Sphinx==4.1.2
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-bibtex==2.3.0
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==2.0.0
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.5
subprocess32==3.5.4
sympy==1.8
tabulate==0.8.9
tensorboard==2.6.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.0
terminado==0.11.1
testpath==0.5.0
threadpoolctl==2.2.0
tokenizers==0.10.3
toml==0.10.2
torch==1.9.0
torch-stft==0.1.4
torchmetrics==0.5.0
torchvision==0.10.0
tornado==6.1
tqdm==4.62.2
traitlets==5.0.5
transformers==4.8.1
typed-ast==1.4.3
typer==0.3.2
typing-extensions==3.10.0.0
Unidecode==1.2.0
urllib3==1.26.6
wandb==0.12.0
wcwidth==0.2.5
webdataset==0.1.62
webencodings==0.5.1
Werkzeug==2.0.1
wget==3.2
widgetsnbextension==3.5.1
wordninja==2.0.0
wrapt==1.12.1
yarg==0.1.9
yarl==1.6.3
youtokentome==1.0.6
